odl/contrib/pytorch/layer.py

```diff
@@ -0,0 +1,245 @@
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+
+"""Utilities for converting ODL operators to pytorch layers.
+
+This requires the ``torch`` module to be available through the
+``pytorch`` package available throught ``pip`` or ``conda``.
+"""
+
+import numpy as np
+import torch
+
+__all__ = ('TorchOperator',)
+
+# TODO: ProductSpaceOperator as multi-input, multi-output construction?
+# TODO: fix issue with stride-0 array
+
+
+class TorchOperator(torch.autograd.Function):
+
+    """Wrap an ODL operator as a Torch autograd Function."""
+
+    def __init__(self, operator):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        operator : `Operator`
+            The ODL operator to be wrapped. For gradient computations to
+            work, ``operator.derivative(x).adjoint`` must be implemented.
+
+        Examples
+        --------
+        >>> space = odl.rn(3)
+        >>> matrix = np.array([[1, 0, 1],
+        ...                    [0, 1, 1]], dtype=float)
+        >>> op = odl.MatrixOperator(matrix, domain=space)
+        >>> matrix_op = TorchOperator(op)
+        >>> matrix_op.operator is op
+        True
+        """
+        self.operator = operator
+
+    def forward(self, input):
+        """Evaluate forward pass on the input.
+
+        Parameters
+        ----------
+        input : `torch.tensor._TensorBase`
+            Point at which to evaluate the operator.
+
+        Returns
+        -------
+        result : `torch.autograd.variable.Variable`
+            Variable holding the result of the evaluation.
+
+        Examples
+        --------
+        Perform a matrix multiplication:
+
+        >>> space = odl.rn(3)
+        >>> matrix = np.array([[1, 0, 1],
+        ...                    [0, 1, 1]], dtype=float)
+        >>> op = odl.MatrixOperator(matrix, domain=space)
+        >>> matrix_op = TorchOperator(op)
+        >>> x_elem = space.element([1, 2, 3])
+        >>> x = torch.from_numpy(x_elem.asarray())
+        >>> x_var = torch.autograd.Variable(x)
+        >>> matrix_op(x_var)
+        Variable containing:
+         4
+         5
+        [torch.DoubleTensor of size 2]
+
+        Evaluate a functional, i.e., an operator with scalar output:
+
+        >>> space = odl.rn(3)
+        >>> functional = odl.solvers.L2NormSquared(space)
+        >>> func_op = TorchOperator(functional)
+        >>> x_elem = space.element([1, 2, 3])
+        >>> x = torch.from_numpy(x_elem.asarray())
+        >>> x_var = torch.autograd.Variable(x)
+        >>> func_op(x_var)
+        Variable containing:
+         14
+        [torch.DoubleTensor of size 1]
+        """
+        if not self.operator.is_linear:
+            # Only needed for nonlinear operators
+            self.save_for_backward(input)
+
+        input_arr = input.numpy()
+        if all(s == 0 for s in input_arr.strides):
```

Comment:
I guess this will be a problem if there are at least 2 zero strides?

Reply:
Obviously, it should be `any`